File: excelium-xls/src/main/java/excelium/xls/ExcelReader.java
/*
 * MIT License
 *
 * Copyright (c) 2018 Excelium
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package excelium.xls;

import excelium.common.ss.CellLocation;
import excelium.common.ss.DateUtil;
import excelium.common.ss.RangeLocation;
import excelium.core.reader.AbstractTestReader;
import org.apache.commons.lang3.StringUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.apache.poi.hssf.usermodel.HSSFWorkbook;
import org.apache.poi.ss.usermodel.*;
import org.apache.poi.xssf.usermodel.XSSFRichTextString;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.file.Paths;
import java.util.*;

import static excelium.common.NumberUtil.getIntValue;
import static excelium.common.NumberUtil.getNumericValue;

/**
 * Reads Microsoft Excel files.
 *
 * @author PhungDucKien
 * @since 2018.04.10
 */
public class ExcelReader extends AbstractTestReader<Workbook, Sheet> {

    /**
     * Logger
     */
    private static final Logger LOG = LogManager.getLogger();

    /**
     * File path
     */
    private String filePath;

    /**
     * Instantiates a new reader for a specific Excel file.
     *
     * @param filePath Path of file to read
     * @throws IOException if any IOException occurs
     */
    ExcelReader(String filePath) throws IOException {
        File file = Paths.get(filePath).toFile();

        this.filePath = filePath;
        if (file.getName().endsWith(".xlsx")) {
            this.workbook = new XSSFWorkbook(new FileInputStream(file));
        } else {
            this.workbook = new HSSFWorkbook(new FileInputStream(file));
        }
    }

    @Override
    public Workbook getWorkbook() throws IOException {
        return workbook;
    }

    @Override
    public String getWorkbookName() throws IOException {
        return filePath;
    }

    @Override
    public List<Sheet> listSheets() throws IOException {
        List<Sheet> sheetList = new ArrayList<>();
        Iterator<Sheet> sheetIterator = this.workbook.sheetIterator();
        while (sheetIterator.hasNext()) {
            Sheet sheet = sheetIterator.next();
            sheetList.add(sheet);
        }
        return sheetList;
    }

    @Override
    public String getSheetName(Sheet sheet) {
        return sheet.getSheetName();
    }

    @Override
    public Map<Object, String> batchFindFirstOccurrence(List<Object> values, Sheet sheet) throws IOException {
        Map<Object, String> cellLocations = new HashMap<>();
        for (Object value : values) {
            if (StringUtils.isNotBlank(findFirstOccurrence(value, sheet))) {
                cellLocations.put(value, findFirstOccurrence(value, sheet));
            }
        }
        return cellLocations;
    }

    @Override
    public String findFirstOccurrence(Object value, Sheet sheet) {
        int firstRow = sheet.getFirstRowNum() < 0 ? 0 : sheet.getFirstRowNum();
        int lastRow = sheet.getLastRowNum() < 0 ? 0 : sheet.getLastRowNum();

        for (int i = firstRow; i <= lastRow; i++) {
            Row r;
            try {
                r = sheet.getRow(i);
            } catch (RuntimeException e) {
                throw e;
            }
            if (r != null) {
                int firstCol = r.getFirstCellNum() < 0 ? 0 : r.getFirstCellNum();
                int lastCol = r.getLastCellNum() < 0 ? 0 : r.getLastCellNum();

                for (int j = firstCol; j <= lastCol; j++) {
                    Object cellValue = getCellValue(sheet, i, j);
                    if (cellValue != null && cellValue.equals(value)) {
                        CellLocation cellLocation = new CellLocation(sheet.getSheetName(), i, j, false, false);
                        return cellLocation.formatAsString();
                    }
                }
            }
        }
        return null;
    }

    @Override
    public Map<String, List<List<Object>>> batchGetRangeCellValues(String... ranges) throws IOException {
        Map<String, List<List<Object>>> rangeCellValues = new HashMap<>();
        for (String range : ranges) {
            RangeLocation rangeLocation = new RangeLocation(range);
            String sheetName = rangeLocation.getFirstCell().getSheetName();
            Sheet sheet = workbook.getSheet(sheetName);
            List<List<Object>> values = new ArrayList<>();
            for (int r = rangeLocation.getFirstCell().getRow(); r <= rangeLocation.getLastCell().getRow(); r++) {
                List<Object> rowValues = new ArrayList<>();
                for (int c = rangeLocation.getFirstCell().getCol(); c <= rangeLocation.getLastCell().getCol(); c++) {
                    rowValues.add(getCellValue(sheet, r, c));
                }
                values.add(rowValues);
            }
            rangeCellValues.put(rangeLocation.formatAsString(), values);
        }
        return rangeCellValues;
    }

    /**
     * Get value of a cell at specific location in a given sheet
     *
     * @param sheet Sheet
     * @param row   Row index
     * @param col   Column index
     * @return Cell value
     */
    private Object getCellValue(Sheet sheet, int row, int col) {
        Row r;
        try {
            r = sheet.getRow(row);
        } catch (RuntimeException e) {
            LOG.error("Error at " + row + "," + col);
            throw e;
        }

        if (r != null) {
            Cell c;
            try {
                c = r.getCell(col);
            } catch (RuntimeException e) {
                LOG.error("Error at " + row + "," + col);
                throw e;
            }
            if (c != null) {
                CellStyle cellStyle = c.getCellStyle();
                Font cellFont = sheet.getWorkbook().getFontAt(cellStyle.getFontIndex());
                if (cellFont != null && cellFont.getStrikeout()) {
                    return null;
                }

                switch (c.getCellTypeEnum()) {
                    case STRING:
                        RichTextString richText = c.getRichStringCellValue();
                        if (richText != null) {
                            XSSFRichTextString xssfRichText = (XSSFRichTextString) richText;
                            int runs = xssfRichText.numFormattingRuns();
                            if (runs > 0) {
                                StringBuilder plainText = new StringBuilder();
                                for (int i = 0; i < runs; i++) {
                                    Font indexFont = xssfRichText.getFontOfFormattingRun(i);
                                    if (indexFont == null) {
                                        indexFont = cellFont;
                                    }
                                    if (indexFont == null || !indexFont.getStrikeout()) {
                                        plainText.append(xssfRichText.getString().substring(xssfRichText.getIndexOfFormattingRun(i),
                                                xssfRichText.getIndexOfFormattingRun(i)
                                                        + xssfRichText.getLengthOfFormattingRun(i)));
                                    }
                                }
                                if (plainText.length() > 0) {
                                    return plainText.toString();
                                } else {
                                    return null;
                                }
                            }
                        }
                        return c.getStringCellValue();
                    case NUMERIC:
                        return getNumericCellValue(c);
                    case BOOLEAN:
                        return c.getBooleanCellValue();
                    case FORMULA:
                        switch (c.getCachedFormulaResultTypeEnum()) {
                            case STRING:
                                return c.getStringCellValue();
                            case NUMERIC:
                                return getNumericCellValue(c);
                            case BOOLEAN:
                                return c.getBooleanCellValue();
                        }
                }
            }
        }
        return null;
    }

    /**
     * Get numeric cell value.
     * This method converts cell value to appropriate data type using cell data format.
     *
     * @param cell Cell object
     * @return numeric cell value
     */
    private Object getNumericCellValue(Cell cell) {
        String format = cell.getCellStyle().getDataFormatString();
        if (StringUtils.isBlank(format) || "General".equals(format)) {
            return getNumericValue(cell.getNumericCellValue());
        } else if (StringUtils.equalsAnyIgnoreCase(format, "TEXT", "@")) {
            return String.valueOf(getNumericValue(cell.getNumericCellValue()));
        } else if (DateUtil.isADateFormat(format)) {
            return cell.getDateCellValue();
        } else if (StringUtils.containsAny(format, ".0", ".#", ".?")) {
            return cell.getNumericCellValue();
        } else {
            return getIntValue(cell.getNumericCellValue());
        }
    }
}

File: packages/xls/src/ExcelReader.ts
/*
 * Licensed to the Software Freedom Conservancy (SFC) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The SFC licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { CellLocation, NumberUtil, RangeLocation, StringUtil } from '@excelium/common';
import { AbstractTestReader } from '@excelium/core';
import {
  Cell,
  CellErrorValue,
  CellFormulaValue,
  CellHyperlinkValue,
  CellRichTextValue,
  CellSharedFormulaValue,
  FormulaType,
  Row,
  ValueType,
  Workbook,
  Worksheet,
} from 'exceljs';

/**
 * Reads Microsoft Excel files.
 *
 * @author PhungDucKien
 * @since 2018.04.10
 */
export default class ExcelReader extends AbstractTestReader<Workbook, Worksheet> {
  /**
   * File path
   */
  private filePath: string;

  /**
   * Instantiates a new reader for a specific Excel file.
   *
   * @param filePath Path of file to read
   */
  constructor(filePath: string) {
    super();
    this.filePath = filePath;
  }

  public getWorkbook(): Promise<Workbook> {
    const workbook = new Workbook();
    return workbook.xlsx.readFile(this.filePath);
  }

  public getWorkbookName(_: Workbook): string {
    return this.filePath;
  }

  public listSheets(workbook: Workbook): Worksheet[] {
    return workbook.worksheets;
  }

  public getSheetName(sheet: Worksheet): string {
    return sheet.name;
  }

  public async batchFindFirstOccurrenceSheet(values: any[], sheet: Worksheet): Promise<Map<string, string>> {
    const cellLocations = new Map<string, string>();
    for (const value of values) {
      const cellLocation = await this.findFirstOccurrenceSheet(value, sheet);
      if (cellLocation && StringUtil.isNotBlank(cellLocation)) {
        cellLocations.set(value, cellLocation);
      }
    }
    return cellLocations;
  }

  public async batchGetRangeCellValues(ranges: string[], workbook: Workbook): Promise<Map<string, any[][]>> {
    const rangeCellValues = new Map<string, any[][]>();
    for (const range of ranges) {
      const rangeLocation = new RangeLocation({ reference: range });
      const sheetName = rangeLocation.getFirstCell().getSheetName();
      const sheet = workbook.getWorksheet(sheetName!);
      const values = [] as any[][];
      for (let r = rangeLocation.getFirstCell().getRow(); r <= rangeLocation.getLastCell().getRow(); r++) {
        const rowValues = [] as any[];
        for (let c = rangeLocation.getFirstCell().getCol(); c <= rangeLocation.getLastCell().getCol(); c++) {
          rowValues.push(this._getCellValue(sheet, r + 1, c + 1));
        }
        values.push(rowValues);
      }
      rangeCellValues.set(rangeLocation.formatAsString(), values);
    }
    return rangeCellValues;
  }

  public async findFirstOccurrenceSheet(value: any, sheet: Worksheet): Promise<string | null> {
    const firstRow = 0;
    const lastRow = sheet.rowCount;

    for (let i = firstRow; i < lastRow; i++) {
      let r: Row;
      try {
        r = sheet.getRow(i);
      } catch (e) {
        throw e;
      }
      if (r != null) {
        const firstCol = 0;
        const lastCol = r.actualCellCount;

        for (let j = firstCol; j < lastCol; j++) {
          const cellValue = this._getCellValue(sheet, i + 1, j + 1);
          if (cellValue != null && cellValue === value) {
            const cellLocation = new CellLocation({ sheetName: sheet.name, row: i, col: j });
            return cellLocation.formatAsString();
          }
        }
      }
    }
    return null;
  }

  /**
   * Get value of a cell at specific location in a given sheet
   *
   * @param sheet Sheet
   * @param row   Row index
   * @param col   Column index
   * @return Cell value
   */
  private _getCellValue(sheet: Worksheet, row: number, col: number): any {
    let r: Row;
    try {
      r = sheet.getRow(row);
    } catch (e) {
      throw e;
    }
    if (r != null) {
      let c: Cell;
      try {
        c = r.getCell(col);
      } catch (e) {
        throw e;
      }
      if (c != null) {
        const cellFont = c.font;
        if (cellFont && cellFont.strike) {
          return null;
        }

        switch (c.type) {
          case ValueType.RichText:
            const richText = c.value as CellRichTextValue;
            const runs = richText.richText.length;
            if (runs > 0) {
              let plainText = '';
              for (let i = 0; i < runs; i++) {
                const text = richText.richText[i];
                let indexFont = text.font;
                if (!indexFont) {
                  indexFont = cellFont;
                }
                if (!indexFont || !indexFont.strike) {
                  plainText += text.text;
                }
              }
              if (plainText.length > 0) {
                return plainText;
              } else {
                return null;
              }
            }
            return null;
          case ValueType.Error:
            return (c.value as CellErrorValue).error;
          case ValueType.Hyperlink:
            return (c.value as CellHyperlinkValue).text;
          case ValueType.SharedString:
          case ValueType.String:
          case ValueType.Date:
          case ValueType.Boolean:
            return c.value;
          case ValueType.Number:
            return this.getNumericFormatValue(c.value, c.numFmt);
          case ValueType.Formula:
            let result: number | string | Date | { error: CellErrorValue } | undefined;
            if (c.formulaType === FormulaType.Master) {
              result = (c.value as CellFormulaValue).result;
            } else if (c.formulaType === FormulaType.Shared) {
              result = (c.value as CellSharedFormulaValue).result;
            }
            switch (c.effectiveType) {
              case ValueType.Number:
                return this.getNumericFormatValue(result != null ? result : 0, c.numFmt);
              case ValueType.String:
                return result != null ? result : '';
              case ValueType.Boolean:
                return result != null ? result : false;
            }
            return result || null;
        }
      }
      return null;
    }
  }

  /**
   * Get numeric format value.
   * This method converts cell value to appropriate data type using cell data format.
   *
   * @param value  Cell value
   * @param format Cell format
   * @return numeric format value
   */
  private getNumericFormatValue(value: any, format: string): any {
    if (StringUtil.isBlank(format) || format === 'General') {
      return value;
    } else if (StringUtil.equalsAnyIgnoreCase(format, ['TEXT', '@'])) {
      return value.toString();
    } else if (StringUtil.containsAny(format, ['.0', '.#', '.?'])) {
      return value;
    } else {
      return NumberUtil.getIntValue(value);
    }
  }
}
